Migrate `codegen_ssa` to diagnostics structs - [Part 2]

JhonnyBillM opened the pull request:
Completes migrating `link.rs` in `codegen_ssa` module.

_Part 1 - https://github.com/rust-lang/rust/pull/102612_

r? @davidtwco

JhonnyBillM committed:
UPDATE - Complete link.rs migration to new diagnostics infraestructure

JhonnyBillM committed:
ADD - ExtractBundledLibsError. Migrated extract_bundled_libs to translatable diagnostics

JhonnyBillM committed:
FIX - Migrate missing errors in link.rs

davidtwco commented on compiler/rustc_codegen_ssa/src/back/archive.rs:
```diff
@@ -28,32 +29,54 @@ pub trait ArchiveBuilderBuilder {
+            Mmap::map(File::open(rlib).map_err(|e| ExtractBundledLibsError {
+                kind: OpenFile,
+                rlib,
+                error: e.to_string(),
```
I think we can use the `io::Error` in the type directly?

JhonnyBillM replied:
We can't because different operations return different error types, here we have: `std::io::Error`, `object::Error` and `Utf8Error`.

We could accept a `&dyn Display` instead if you think that's better than `String` here.

davidtwco replied:
What about a `&dyn std::error::Error`?

JhonnyBillM replied:
Updated [here](https://github.com/rust-lang/rust/pull/103792/commits/d43977c5fd7bac51bc22aa49bdc5d9cf9db52569) to `Box<dyn Error>` and made it conform to `IntoDiagnosticArg`. 
 
Please let me know if there is a better way to implement this.

davidtwco commented on compiler/rustc_codegen_ssa/src/errors.rs:
```diff
@@ -354,3 +354,194 @@ impl IntoDiagnostic<'_> for LinkingFailed<'_> {
+pub struct OptionGccOnly;
+
+pub struct ExtractBundledLibsError<'a> {
+    pub kind: ExtractBundledLibsErrorKind,
```
Can we just make this an enum and do away with the `ExtractBundleLibsErrorKind` enum?

JhonnyBillM replied:
Yes we can, but I tried to prevent making harder to distinguish each enum case, and avoid some repetition.

I am not sure what's the Rust-idiomatic way to handle when all enum cases share the same parameters, but in Swift we tend to break them into a struct + enum.

This is how it looks, let me know if you think we should still change it only use the enum.

```rust
pub enum ExtractBundledLibsError<'a> {
    #[diag(codegen_ssa_extract_bundled_libs_open_file)]
    OpenFile { rlib: &'a Path, error: String },

    #[diag(codegen_ssa_extract_bundled_libs_mmap_file)]
    MmapFile { rlib: &'a Path, error: String },

    #[diag(codegen_ssa_extract_bundled_libs_parse_archive)]
    ParseArchive { rlib: &'a Path, error: String },

    #[diag(codegen_ssa_extract_bundled_libs_read_entry)]
    ReadEntry { rlib: &'a Path, error: String },

    #[diag(codegen_ssa_extract_bundled_libs_archive_member)]
    ArchiveMember { rlib: &'a Path, error: String },

    #[diag(codegen_ssa_extract_bundled_libs_convert_name)]
    ConvertName { rlib: &'a Path, error: String },
    
    #[diag(codegen_ssa_extract_bundled_libs_write_file)]
    WriteFile { rlib: &'a Path, error: String },
}
```

davidtwco replied:
I think this is fine, there's a bit of duplication but that's okay.

JhonnyBillM replied:
Updated [here](https://github.com/rust-lang/rust/pull/103792/commits/31cdab1df18f0875cedb600d602ca173216980ae)

davidtwco commented on compiler/rustc_codegen_ssa/src/errors.rs:
```diff
@@ -354,3 +354,194 @@ impl IntoDiagnostic<'_> for LinkingFailed<'_> {
+#[derive(Diagnostic)]
+#[diag(codegen_ssa_stripping_debu_info_failed)]
+#[note]
+pub struct StrippingDebuInfoFailed<'a> {
```
typo?

JhonnyBillM replied:
Thanks! Fixed [here](https://github.com/rust-lang/rust/pull/103792/commits/ca584546f4e12d6a289e26ab7b2f9badf39d753a)

davidtwco commented on compiler/rustc_codegen_ssa/src/back/link.rs:
```diff
@@ -1092,21 +1063,21 @@ fn strip_symbols_with_external_utility<'a>(
 fn escape_string(s: &[u8]) -> String {
     match str::from_utf8(s) {
         Ok(s) => s.to_owned(),
+        // FIXME: return a type that can conform to IntoDiagnosticArg
```
I think this is okay, command output won't ever be translated so it is fine?

JhonnyBillM replied:
Makes sense. Addressed [here](https://github.com/rust-lang/rust/pull/103792/commits/ca584546f4e12d6a289e26ab7b2f9badf39d753a)

davidtwco reviewed (changes requested):
This looks really good, left a few minor comments.

JhonnyBillM committed:
UPDATE - address PR Comments

FIX - StrippingDebugInfoFailed typo

DELETE - unneeded FIXME comment

UPDATE - only declare the error with ExtractBundledLibsError as an enum and use the Diagnostic derive macro

JhonnyBillM committed:
UPDATE - accept dyn error and make Box<dyn error> conform to IntoDiagnosticArg

davidtwco commented:
@bors r+

bors commented:
:pushpin: Commit 540c3f94d71879f413a151bc8c83c20c10b386dc has been approved by `davidtwco`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=davidtwco 540c3f94d71879f413a151bc8c83c20c10b386dc -->
<!-- homu: {"type":"Approved","sha":"540c3f94d71879f413a151bc8c83c20c10b386dc","approver":"davidtwco","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
